app/src/main/java/io/github/cmput301f19t19/legendary_fiesta/FirebaseHelper.java

```diff
@@ -0,0 +1,115 @@
+        void onSuccess(T document);
+
+        /**
+         * callback handler for handling failure cases
+         * @param e Exception thrown by the Firebase library
+         */
+        void onFailure(@NonNull Exception e);
+    }
+
+    /**
+     * check if the username has already been taken
+     * @param name the username to check
+     * @param callback callback, called when the query finishes, needs to be of type FirebaseCallback<QuerySnapshot>
+     */
+    public void checkUserExists(String name, final FirebaseCallback<QuerySnapshot> callback) {
+        db.collection("users").whereEqualTo("username", name).get().addOnSuccessListener(new OnSuccessListener<QuerySnapshot>() {
+            @Override
+            public void onSuccess(QuerySnapshot queryDocumentSnapshots) {
+                callback.onSuccess(queryDocumentSnapshots);
+            }
+        }).addOnFailureListener(new OnFailureListener() {
+            @Override
+            public void onFailure(@NonNull Exception e) {
+                callback.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * @param uid UID generated and tracked by Firebase
+     * @param callback callback callback, called when the query finishes, needs to be of type FirebaseCallback<DocumentSnapshot>
+     */
+    public void getUserByUID(String uid, final FirebaseCallback<DocumentSnapshot> callback) {
+        db.collection("users").document(uid).get().addOnSuccessListener(new OnSuccessListener<DocumentSnapshot>() {
+            @Override
+            public void onSuccess(DocumentSnapshot documentSnapshot) {
+                callback.onSuccess(documentSnapshot);
+            }
+        }).addOnFailureListener(new OnFailureListener() {
+            @Override
+            public void onFailure(@NonNull Exception e) {
+                callback.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * add an user to the database
+     * @param user user to be added
+     * @param callback callback, called when the query finishes, needs to be of type FirebaseCallback<DocumentReference>
+     */
+    public void addUser(User user, final FirebaseCallback<DocumentReference> callback) {
+        db.collection("users").document(user.getUid()).set(user).addOnSuccessListener(new OnSuccessListener<Void>() {
+            @Override
+            public void onSuccess(Void aVoid) {
+                callback.onSuccess(null);
+            }
+        }).addOnFailureListener(new OnFailureListener() {
+            @Override
+            public void onFailure(@NonNull Exception e) {
+                callback.onFailure(e);
+            }
+        });
+    }
+
+    /**
+     * add a mood event to the database
+     * @param moodEvent MoodEvent to be added
+     * @param callback callback, called when the query finishes, needs to be of type FirebaseCallback<DocumentReference>
+     */
+    public void addMoodEvent(MoodEvent moodEvent, final FirebaseCallback<DocumentReference> callback) {
```

Comment:
Should we separate mood and user firebase actions across different classes? Maybe they could extend an Abstract FirebaseHelper Class?

Reply:
>Maybe they could extend an Abstract FirebaseHelper Class?

It would be difficult and inefficient. Although the  `FirebaseFirestore` is itself `static`, it cannot be declared to be `static` elsewhere due to how it manages its own states. 

Also, by splitting this helper, you will probably need to re-import or re-define the nested interface `FirebaseCallback`.

>Should we separate mood and user firebase actions across different classes?

Although I think the idea is good, I would like to keep it as-is at least for now.

Reply:
In addition, I am probably going to generify all the `add`, `set` and `delete` operations so that we can avoid duplicate code. And these functions are going to be `private` in this class to hide the implementation details and to be more compliant with the practices.

Reply:
Sounds good, not something urgent or necessary for now anyway!